### PR TITLE
Login and SignUp

### PR DESCRIPTION
**Pull request by SeR4h**
@dennisja Please review this when you get time.Thanks

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "myexpensesapi",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
[the remark this answers is not in the document]
Comment (SeR4h) [reply package.json:7]: Not sure where to start but am working on it

File: routes/login.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
+const express = require('express');
+var jwt = require('jsonwebtoken');
+const bcrypt = require('bcrypt');
+var pool = require('../models/db');
+const login = (app) => {
+    app.post('/login', (request, response, next) => {
+        var username = request.body.username;
+        var password = request.body.password;
+
+        pool.query('SELECT * FROM users WHERE username = ?', username, (error, result) => {
```
[the remark this answers is not in the document]
Comment (SeR4h) [reply routes/login.js:12]: It returns username doesnot exist
I've heard about Sql injections though

**Comment by SeR4h** [comment]
@dennisja please review the corrections i've made. 
Thankyou

File: models/db.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+/* eslint-disable linebreak-style */
+/* eslint-disable indent */
+/* eslint-disable linebreak-style */
+require('dotenv').config();
```
Comment (dennisja) [inline models/db.js:4]: You only need to do this once. For example, it's better to extract your env variables from one place e.g you can create a file say `config.js` where you get the values from environment variables and export them it cabe something like
```js
require('dotenv').config();

const { host, password, someOtherEnvVar} = process.env

module.exports = { host, password, someOtherEnvVar }
```

And in your other parts of the application you pick the values from `configs` such that you don't feel like you need to call `require('dotenv').config();` again

File: routes/login.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
+/* eslint-disable linebreak-style */
+/* eslint-disable indent */
+/* eslint-disable linebreak-style */
+const jwt = require('jsonwebtoken');
+const bcrypt = require('bcrypt');
+require('dotenv').config();
```
Comment (dennisja) [inline routes/login.js:6]: Let's have this in one place

File: routes/login.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+const pool = require('../models/db');
+
+const login = (app) => {
+    app.post('/login', (request, response) => {
+        const { username, password } = request.body;
+        pool.query('SELECT * FROM users WHERE username = ?', username, (error, result) => {
+            if (result.length > 0) {
```
Comment (dennisja) [inline routes/login.js:13]: Your code is too nested, we will talk about how to make it less nested. Nesting reduces readability. Do you know anything about [`Promises`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) and [`async...await`](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Async_await) yet?

File: routes/login.js (hunk 4 above)
Comment (dennisja) [inline routes/login.js:6]: You won't need to re-import this

File: server.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/* eslint-disable linebreak-style */
+/* eslint-disable no-console */
+/* eslint-disable indent */
+/* eslint-disable linebreak-style */
+const express = require('express');
+const bodyParser = require('body-parser');
+const verifyToken = require('./verifyToken');
+require('dotenv').config();
```
Comment (dennisja) [inline server.js:8]: You won't need to do this here

File: server.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+/* eslint-disable linebreak-style */
```
Comment (dennisja) [inline server.js:1]: Use `eslintrc.js` to configure or turn off the rules

File: routes/login.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
+/* eslint-disable linebreak-style */
```
Comment (dennisja) [inline routes/login.js:1]: User `eslintrc.js` to configure or turn off the rules

File: routes/signup.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
+/* eslint-disable linebreak-style */
```
Comment (dennisja) [inline routes/signup.js:1]: User `eslintrc.js` to configure or turn off the rules

File: routes/signup.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+/* eslint-disable indent */
+/* eslint-disable linebreak-style */
+const bcrypt = require('bcrypt');
+const pool = require('../models/db');
+
+const signup = (app) => {
+    app.post('/signup', async (request, response) => {
+        const { password } = request.body;
+        bcrypt.hash(password, 10, (err, hash) => {
+            const user = {
```
Comment (dennisja) [inline routes/signup.js:11]: You will need to do some validation, but this can be done later in a separate PR

File: verifyToken.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
+/* eslint-disable linebreak-style */
```
Comment (dennisja) [inline verifyToken.js:1]: Override or turnoff lint rules in the config file unless its a one-off thing

File: verifyToken.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+/* eslint-disable indent */
+/* eslint-disable linebreak-style */
+const jwt = require('jsonwebtoken');
+require('dotenv').config();
+
+module.exports.verifyToken = (request, response, next) => {
+    const bearerHeader = request.headers.authorization;
+    if (typeof bearerHeader !== 'undefined') {
+        const bearer = bearerHeader.split(' ');
```
Comment (dennisja) [inline verifyToken.js:10]: You could  as well use array destructuring to get the bearer token i.e
```js
const [, bearerToken] = bearerHeader.split(' ');
```
You can read more about array destructuring [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) or [here](https://www.taniarascia.com/understanding-destructuring-rest-spread/)

File: expensesdb.sql [hunk 13]
```diff
@@ -0,0 +1,114 @@
+-- phpMyAdmin SQL Dump
```
[the remark this answers is not in the document]
Comment (SeR4h) [reply expensesdb.sql:1]: Noted

**Comment by SeR4h** [comment]
@dennisja please review final modifications.Thnx☺

**Pull request merged**